### PR TITLE
Tax reports not showing correct totals

### LINKED ISSUE #42682: Tax reports not showing correct totals
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

On the Analytics > Taxes report, the actual total of the numbers in each column (Total Tax, Order Tax) is not the same as the total shown at the bottom of the chart and on the graph. This only seems to happen when a custom date range is indicated. Also, while I can duplicate it on my test site, it does not happen for all date ranges.

### Expected behavior

Total should match the total of numbers in the columns.

### Actual behavior

For example, on my test site, I have several tax rates set up that are specific to different postal codes in New York State.

On the Analytics > Taxes report, I can specify a custom date range of September 20-21 2023. On my test site, this matches to 4 orders, two from the 20th, 2 from the 21st.

The total Order Tax and Shipping Tax reported on the graph, and at the top of the screen, are correct for those 4 orders. However, in the line chart, only 3 orders are reported as matching and the numbers in the columns don't match the total.

### Steps to reproduce

1. Set up tax rates for various specific areas (on my test site, I have different zip codes in NY State; the original reporting customer uses different tax rates for different European countries).
2. Make a few orders across a few days.
3. Run the Analytics > Taxes report and compare line by line report in the chart at the bottom of the page numbers, with the total reported, when using a custom date range.

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://numberone.mystagingwebsite.com
Site address (URL): https://numberone.mystagingwebsite.com
WC Version: 8.3.1
REST API Version: ✔ 8.3.1
WC Blocks Version: ✔ 11.4.9
Action Scheduler Version: ✔ 3.6.4
Log Directory Writable: ✔
WP Version: 6.4.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 8.2.13
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 8.4.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.4.25-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.3.1
WC Database Prefix: wp_
Total Database Size: 126.24MB
Database Data Size: 80.39MB
Database Index Size: 45.85MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.16MB + Index: 0.08MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 1.52MB + Index: 1.91MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 30.08MB + Index: 20.22MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 12.02MB + Index: 12.03MB + Engine InnoDB
wp_automatewoo_abandoned_carts: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_automatewoo_customers: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_automatewoo_customer_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_guests: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_automatewoo_guest_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_automatewoo_log_meta: Data: 0.13MB + Index: 0.14MB + Engine InnoDB
wp_automatewoo_queue: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_automatewoo_queue_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_referrals: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_automatewoo_referral_advocate_keys: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_automatewoo_referral_invites: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_ce4wp_abandoned_checkout: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_ce4wp_contacts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.23MB + Index: 0.16MB + Engine InnoDB
wp_gla_attribute_mapping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.19MB + Index: 0.14MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_jetpack_sync_queue: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_custom_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_dynamic_segment_filters: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_feature_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_forms: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_migrations: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_newsletter_links: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_newsletter_option: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_option_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_segment: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_templates: Data: 2.52MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_scheduled_tasks: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_scheduled_task_subscribers: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_segments: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_sending_queues: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_settings: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_bounces: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_statistics_clicks: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_statistics_opens: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_mailpoet_statistics_unsubscribes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_woocommerce_purchases: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_mailpoet_stats_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_subscribers: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_mailpoet_subscriber_custom_field: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_ips: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_segment: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_subscriber_tag: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_tags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_user_agents: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_user_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.05MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 4.52MB + Index: 6.84MB + Engine InnoDB
wp_posts: Data: 0.47MB + Index: 0.31MB + Engine InnoDB
wp_redirection_404: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_redirection_groups: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_redirection_items: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_redirection_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.13MB + Index: 0.06MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.08MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_bundle_lookup: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_composite_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.09MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_bundled_itemmeta: Data: 0.08MB + Index: 0.09MB + Engine InnoDB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_exported_csv_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_gc_activity: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_gc_cards: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_gc_cardsmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_flat_rate_boxes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_table_rates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpml_mails: Data: 23.52MB + Index: 0.00MB + Engine InnoDB
wp_yoast_indexable: Data: 0.14MB + Index: 0.09MB + Engine InnoDB
wp_yoast_indexable_hierarchy: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_yoast_migrations: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_yoast_primary_term: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_yoast_seo_links: Data: 0.02MB + Index: 0.03MB + Engine InnoDB

### Post Type Counts ###

acf-field: 1
acf-field-group: 1
attachment: 185
aw_workflow: 10
enr_email_template: 1
global_product_addon: 2
mailpoet_page: 1
nav_menu_item: 6
page: 14
post: 3
product: 155
product_variation: 174
revision: 37
shop_coupon: 6
shop_order: 648
shop_order_refund: 4
shop_subscription: 64
wcpf_item: 8
wcpf_project: 2
wc_membership_plan: 5
wc_user_membership: 13
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (35) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3
AutomateWoo - Refer A Friend Add-on: by WooCommerce – 2.7.12
AutomateWoo: by WooCommerce – 6.0.9
Code Snippets: by Code Snippets Pro – 3.6.2
Debug Bar: by wordpressdotorg – 1.1.4
Facebook for WooCommerce: by Facebook – 3.1.5
Google Listings and Ads: by WooCommerce – 2.5.13
Jetpack: by Automattic – 12.9-beta
Loco Translate: by Tim Whitlock – 2.6.6
Redirection: by John Godley – 5.3.10
WooCommerce All Products For Subscriptions: by WooCommerce – 4.1.4
WooCommerce Composite Products: by WooCommerce – 8.10.4
WooCommerce Distance Rate Shipping: by WooCommerce – 1.3.3
WooCommerce Gift Cards: by WooCommerce – 1.16.6
WooCommerce Google Analytics Integration: by WooCommerce – 1.8.10
WooCommerce Memberships: by SkyVerge – 1.25.2
WooCommerce Name Your Price: by Kathy Darling – 3.5.8 (update to version 3.5.9 is available)
WooPayments: by Automattic – 6.9.0
WooCommerce Pre-Orders: by WooCommerce – 2.0.6
WooCommerce Product Add-ons: by WooCommerce – 6.4.7 (update to version 6.5.1 is available)
WooCommerce Product Bundles: by WooCommerce – 6.22.5
WooCommerce Product CSV Import Suite: by WooCommerce – 1.10.66
Product Filters for WooCommerce: by WooCommerce – 1.4.16
WooCommerce Shipping & Tax: by WooCommerce – 2.4.2
WooCommerce Shipment Tracking: by WooCommerce – 2.4.1
WooCommerce Flat Rate Box Shipping: by WooCommerce – 2.2.4
WooCommerce UPS Shipping: by WooCommerce – 3.5.7
WooCommerce - ShipStation Integration: by WooCommerce – 4.3.9
Woo Subscriptions: by WooCommerce – 5.7.0
WooCommerce Table Rate Shipping: by WooCommerce – 3.1.6
WooCommerce Variation Swatches and Photos: by Element Stark – 3.1.7
WooCommerce: by Automattic – 8.3.1
Yoast SEO: by Team Yoast – 21.6
WP Crontrol: by John Blackbourn & crontributors – 1.16.1
WP Mail Logging: by WP Mail Logging Team – 1.12.0

### Inactive Plugins (12) ###

Advanced Custom Fields: by WP Engine – 6.1.7
Advanced Product Fields for WooCommerce: by StudioWombat – 1.6.2
Cart Upsell for WooCommerce: by SKROTRON – 1.0.8
Custom Post Type UI: by WebDevStudios – 1.13.6
Discounts for WooCommerce Subscriptions: by Meow Crew – 2.4.1 (update to version 3.1.0 is available)
Enhancer for WooCommerce Subscriptions: by Flintop – 4.2.0
Extra Fees for WooCommerce: by Flintop – 4.8.0
MailPoet: by MailPoet – 4.35.1
WooCommerce Customer/Order/Coupon CSV Import Suite: by SkyVerge – 3.11.0 (update to version 3.12.0 is available)
WooCommerce Customer/Order/Coupon Export: by SkyVerge – 5.4.1 (update to version 5.5.0 is available)
WooCommerce Quick View: by Themesquad – 1.7.0
WooCommerce Subscriptions Gifting: by WooCommerce – 2.7.0

### Dropin Plugins (2) ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
exclude_from_catalog (exclude_from_catalog)
exclude_from_search (exclude_from_search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)
visible (visible)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: #767 - /terms-and-conditions/

### Theme ###

Name: Storefront
Version: 4.3.0 (update to version 4.5.3 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### WooPayments ###

Version: 6.9.0
Connected to WPCOM: Yes
WPCOM Blog ID: 210717503
Account ID: acct_1Ljq8PFtdYohmYFC
Payment Gateway: Enabled
Test Mode: Enabled
UPE: Enabled (deferred_intent)
UPE Enabled APMs: card
WooPay: Not eligible
Apple Pay / Google Pay: Enabled (product,cart,checkout)
Fraud Protection Level: basic
Multi-currency: Enabled
Public Key Encryption: Disabled
Auth and Capture: Enabled
Documents: Disabled
Logging: Enabled

### Memberships ###

Restriction mode:: Hide content only
Public content:: –
Excluded from discounts:: –
Membership plans:: 5
White Text Membership:: Members: 0Access method: Product(s) purchaseAccess length: Unlimited Subscription: No
Free Membership:: Members: 4Access method: Manual assignment onlyAccess length: Unlimited Subscription: No
Basic membership:: Members: 1Access method: Product(s) purchaseAccess length: Specific length (60 days)Subscription: No
Fruit Basket Delivery:: Members: 4Access method: Product(s) purchaseAccess length: Unlimited Subscription: YesInstallment plan: No No access-granting products are purchasable
VIP Membership:: Members: 4Access method: Product(s) purchaseAccess length: Specific length (6 months)Subscription: YesInstallment plan: No

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: https://numberone.mystagingwebsite.com
Subscriptions-core Library Version: 6.5.0
Subscription Statuses: wc-active: 41
wc-expired: 5
wc-pending-cancel: 2
wc-pending: 2
wc-on-hold: 7
wc-cancelled: 7

WooCommerce Account Connected: ✔ Yes
Active Product Key: ❌ No
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Store Setup ###

Country / State: United States (US) — New York

### Subscriptions by Payment Gateway ###

Check payments: wc-on-hold: 1
WooPayments: wc-active: 39
wc-cancelled: 7
wc-expired: 5
wc-on-hold: 4
wc-pending-cancel: 2

### Payment Gateway Support ###

WooPayments: products
refunds
multiple_subscriptions
subscription_cancellation
subscription_payment_method_change_admin
subscription_payment_method_change_customer
subscription_payment_method_change
subscription_reactivation
subscription_suspension
subscriptions
subscription_amount_changes
subscription_date_changes
tokenization
add_payment_method

Check payments: products
Cash on delivery: products

### Composite Products ###

Database Version: 8.10.4
Loopback Test: ✔
Template Overrides: –

### Gift Cards ###

Database Version: 1.16.6
Loopback Test: ✔
Task Queueing Test: ✔

### Product Bundles ###

Database Version: 6.22.5
Loopback Test: ✔
Template Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-virtual-downloadable
product-external-affiliate
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-12-12 19:51:29 -05:00
Options: ✔
Notes: 158
Onboarding: completed

### All Products for WooCommerce Subscriptions ###

Template Overrides: –

### Action Scheduler ###

Canceled: 27
Oldest: 2023-11-12 03:02:13 -0500
Newest: 2023-12-06 12:12:43 -0500

Complete: 49,790
Oldest: 2023-11-10 15:17:19 -0500
Newest: 2023-12-11 15:14:38 -0500

Failed: 122
Oldest: 2022-10-18 12:07:43 -0400
Newest: 2023-12-10 17:22:28 -0500

Pending: 95
Oldest: 2023-12-11 15:16:38 -0500
Newest: 2024-12-22 02:00:00 -0500

### Name Your Price ###

Version: 
Database version: 
Template overrides: –

### Status report information ###

Generated at: 2023-12-11 15:16:38 -05:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by mdperez86**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #42682

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create two standard rates from `/wp-admin/admin.php?page=wc-settings&tab=tax&section=standard`
2. Create two orders and assign each standard rate you create in point 1 to the orders: rate 1 to the order 1 and rate 2 to the order 2.
3. Save both orders as `Completed`
4. In the `Analytics > Taxes` report the totals should be correctly calculated based on the items shown in the table image
5. After deleting one standard rate created in point 1, it should still be shown in the table and the totals should not be affected while the table should looks like in point 4

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php [hunk 1]
```diff
@@ -61,18 +61,19 @@ class DataStore extends ReportsDataStore implements DataStoreInterface {
...
-			'name'         => 'tax_rate_name as name',
-			'tax_rate'     => 'tax_rate',
-			'country'      => 'tax_rate_country as country',
-			'state'        => 'tax_rate_state as state',
-			'priority'     => 'tax_rate_priority as priority',
+			'name'         => "SUBSTRING_INDEX(SUBSTRING_INDEX({$wpdb->prefix}woocommerce_order_items.order_item_name,'-',-2), '-', 1) as name",
+			'tax_rate'     => "MAX(CAST({$wpdb->prefix}woocommerce_order_itemmeta.meta_value AS DECIMAL(7,4))) as tax_rate",
+			'country'      => "SUBSTRING_INDEX({$wpdb->prefix}woocommerce_order_items.order_item_name,'-',1) as country",
+			'state'        => "SUBSTRING_INDEX(SUBSTRING_INDEX({$wpdb->prefix}woocommerce_order_items.order_item_name,'-',-3), '-', 1) as state",
+			'priority'     => "SUBSTRING_INDEX({$wpdb->prefix}woocommerce_order_items.order_item_name,'-',-1) as priority",
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php:81]: Is there any way we can avoid these `SUBSTRING_INDEX`?

File: plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php [hunk 2]
```diff
@@ -61,18 +61,28 @@ class DataStore extends ReportsDataStore implements DataStoreInterface {
...
+		$table_name = self::get_db_table_name();
+
+		// Using wp_woocommerce_tax_rates table limits the result to only the existing tax rates and
+		// omit the hitorical records which differs from the porpuse of wp_wc_order_tax_lookup table.
+		// So in order to get the same data present in wp_woocommerce_tax_rates without braking the
+		// API contract the values are now get from wp_woocommerce_order_items and wp_woocommerce_order_itemmeta.
+		// And given that country, state and priority are not separate columns within the woocommerce_order_items
+		// a split to order_item_name column value is required to separate those values. This is not ideal,
+		// but given this query is paginated and cached, then it is not a big deal. There is always room for
+		// improvements here.
```
Comment (mattsherman) [inline plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php]:
```suggestion
		// Using wp_woocommerce_tax_rates table limits the result to only the existing tax rates and
		// omits the historical records which differs from the purpose of wp_wc_order_tax_lookup table.
		// So in order to get the same data present in wp_woocommerce_tax_rates without breaking the
		// API contract the values are now retrieved from wp_woocommerce_order_items and wp_woocommerce_order_itemmeta.
		// And given that country, state and priority are not separate columns within the woocommerce_order_items,
		// a split to order_item_name column value is required to separate those values. This is not ideal,
		// but given this query is paginated and cached, then it is not a big deal. There is always room for
		// improvements here.
```

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php [hunk 3]
```diff
@@ -60,20 +78,26 @@ public function test_get_reports() {
...
 
-		$wpdb->insert(
+		$tax_rate_id = $wpdb->insert(
 			$wpdb->prefix . 'woocommerce_tax_rates',
 			array(
 				'tax_rate_id'       => 1,
 				'tax_rate'          => '7',
 				'tax_rate_country'  => 'US',
-				'tax_rate_state'    => 'GA',
+				'tax_rate_state'    => 'NY',
```
Comment (mattsherman) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php:87]: Any reason for this change? I mean, I'm partial to NY 😄, but there doesn't seem to be a need for this change.

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php [hunk 4]
```diff
@@ -102,10 +126,9 @@ public function test_get_reports() {
 		$tax_report = reset( $reports );
 
 		$this->assertEquals( 1, $tax_report['tax_rate_id'] );
-		$this->assertEquals( 'TestTax', $tax_report['name'] );
 		$this->assertEquals( 7, $tax_report['tax_rate'] );
 		$this->assertEquals( 'US', $tax_report['country'] );
-		$this->assertEquals( 'GA', $tax_report['state'] );
+		$this->assertEquals( 'NY', $tax_report['state'] );
```
Comment (mattsherman) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php:131]: Any reason for this change? I mean, I'm partial to NY 😄, but there doesn't seem to be a need for this change.

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php [hunk 5]
```diff
@@ -134,7 +157,7 @@ public function test_get_reports_taxes_param() {
 				'tax_rate_id'       => 1,
 				'tax_rate'          => '7',
 				'tax_rate_country'  => 'US',
-				'tax_rate_state'    => 'GA',
+				'tax_rate_state'    => 'NY',
```
Comment (mattsherman) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php:160]: Any reason for this change? I mean, I'm partial to NY 😄, but there doesn't seem to be a need for this change.

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php [hunk 6]
```diff
@@ -190,140 +244,26 @@ public function test_get_reports_taxes_param() {
...
+		$this->assertEquals( 1, $tax_report['orders_count'] );
 
 		$tax_report = next( $reports );
 
 		$this->assertEquals( 1, $tax_report['tax_rate_id'] );
-		$this->assertEquals( 'TestTax', $tax_report['name'] );
 		$this->assertEquals( 7, $tax_report['tax_rate'] );
 		$this->assertEquals( 'US', $tax_report['country'] );
-		$this->assertEquals( 'GA', $tax_report['state'] );
+		$this->assertEquals( 'NY', $tax_report['state'] );
```
Comment (mattsherman) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php:260]: Any reason for this change? I mean, I'm partial to NY 😄, but there doesn't seem to be a need for this change.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Tested and works as expected. Just two minor changes requested...

- Fixes to typos in comment
- Revert of State change from GA to NY in test, unless there is a reason why that change was necessary

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php (hunk 3 above)
Comment (louwie17) [reply plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-taxes.php:87]: There is a reason, as we make use of the `create_order` helper function here that auto sets the billing/shipping address to NY: https://github.com/woocommerce/woocommerce/blob/e05697dfabe8d5ab6019147729d0858238798b8d/plugins/woocommerce/tests/legacy/framework/helpers/class-wc-helper-order.php#L86 
Meaning that this tax rate won't work if we keep it as GA, as you can see by the below screenshot, that it will fail.
Screenshot 2024-04-18 at 2 36 23 PM

**Comment by louwie17** [comment]
Btw I found a bug, just trying to debug what is going wrong.

**Review by louwie17 — APPROVED** [review]
I tested this again and it tested well, although I did manage to create a random edge case in my test site where it had created two rows for the same order in `order_tax_lookup` table, while it only had one tax rate.
But this is not related to this code and very much appeared to be an edge case.

Maybe just update the comments that @mattsherman suggested.

**Review by mattsherman — APPROVED** [review]
Thanks for updating the comment!

**Comment by louwie17** [comment]
@mdperez86 can we merge this?
